core/src/layers/prometheus.rs

```diff
@@ -98,784 +95,235 @@ use crate::*;
-    type Inner = A;
-    type Reader = PrometheusMetricWrapper<A::Reader>;
-    type BlockingReader = PrometheusMetricWrapper<A::BlockingReader>;
-    type Writer = PrometheusMetricWrapper<A::Writer>;
-    type BlockingWriter = PrometheusMetricWrapper<A::BlockingWriter>;
-    type Lister = A::Lister;
-    type BlockingLister = A::BlockingLister;
-
-    fn inner(&self) -> &Self::Inner {
-        &self.inner
-    }
-
-    async fn create_dir(&self, path: &str, args: OpCreateDir) -> Result<RpCreateDir> {
-        let labels = self.stats.generate_metric_label(
-            self.scheme.into_static(),
-            Operation::CreateDir.into_static(),
-            path,
-        );
-
-        self.stats.requests_total.with_label_values(&labels).inc();
-
-        let timer = self
-            .stats
-            .requests_duration_seconds
-            .with_label_values(&labels)
-            .start_timer();
-        let create_res = self.inner.create_dir(path, args).await;
-
-        timer.observe_duration();
-        create_res.map_err(|e| {
-            self.stats
-                .increment_errors_total(Operation::CreateDir, e.kind());
-            e
-        })
-    }
-
-    async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
-        let labels = self.stats.generate_metric_label(
-            self.scheme.into_static(),
-            Operation::Read.into_static(),
-            path,
-        );
-        self.stats.requests_total.with_label_values(&labels).inc();
-
-        let timer = self
-            .stats
-            .requests_duration_seconds
-            .with_label_values(&labels)
-            .start_timer();
-        let res = self.inner.read(path, args).await;
-        timer.observe_duration();
-
-        match res {
-            Ok((rp, r)) => Ok((
-                rp,
-                PrometheusMetricWrapper::new(
-                    r,
-                    Operation::Read,
-                    self.stats.clone(),
-                    self.scheme,
-                    &path.to_string(),
-                ),
-            )),
-            Err(err) => {
-                self.stats
-                    .increment_errors_total(Operation::Read, err.kind());
-                Err(err)
-            }
-        }
-    }
-
-    async fn write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::Writer)> {
-        let labels = self.stats.generate_metric_label(
-            self.scheme.into_static(),
-            Operation::Write.into_static(),
-            path,
-        );
-        self.stats.requests_total.with_label_values(&labels).inc();
-
-        let timer = self
-            .stats
-            .requests_duration_seconds
-            .with_label_values(&labels)
-            .start_timer();
-        let res = self.inner.write(path, args).await;
-        timer.observe_duration();
+        let mut labels = vec![scheme.into_static(), namespace, root, op.into_static()];
```

Comment:
We introduce an additional allocation here. How about expanding the code by writing it twice instead of calling `push`?